draft-ietf-tls-esni.md

```diff
@@ -719,38 +712,55 @@ error code.
-1. Versions in the "supported_versions" extension, excluding TLS 1.2 and
-earlier. Note the ClientHelloOuter MAY include these older versions, while the
-ClientHelloInner MUST omit them.
+The HelloRetryRequest constructed by the backend server is called the
+HelloRetryRequestInner message; and the HelloRetryRequest constructed by the
+client-facing server is called the HelloRetryRequestOuter message.
 
-Future extensions that might change across first and second ClientHello messages
-in response to a HelloRetryRequest MUST have the same value.
+If the client offered ECH in the first ClientHello, then it MUST offer ECH in
+the second. Likewise, if the client did not offer ECH in the first ClientHello,
+then it MUST NOT not offer ECH in the second.
 
-If the server sends a HelloRetryRequest in response to the ClientHello, the
-client sends a second updated ClientHello per the rules in {{RFC8446}}.
-However, at this point, the client does not know whether the server processed
-ClientHelloOuter or ClientHelloInner, and MUST regenerate both values to be
-acceptable. Note: if ClientHelloOuter and ClientHelloInner use different groups
-for their key shares or differ in some other way, then the HelloRetryRequest
-may actually be invalid for one or the other ClientHello, in which case a fresh
-ClientHello MUST be generated, ignoring the instructions in HelloRetryRequest.
-Otherwise, the usual rules for HelloRetryRequest processing apply.
-
-The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
-the second ClientHelloOuter for any referenced extensions. It then encrypts
-the new EncodedClientHelloInner value as a second message with the previous
-HPKE context:
+#### Accepted ECH
+
+As described in {{sever-hrr}, the HelloRetryRequestInner is encrypted by the
+backend server and included as the payload of the "encrypted_client_hello"
+extension of the HelloRetryRequestOuter. The client decrypts the payload using
+the HPKE context it shares with the client-facing server. It begins by exporting
+a key and nonce for AEAD decryption:
+
+~~~
+    hrr_inner_key = context.Export("hrr inner key", Nk)
+    hrr_inner_nonce = context.Export("hrr inner nonce", Nn)
+~~~
+
+Here `Nk` and `Nn` denote, respectively, the key and nonce length of the AEAD
+algorithm associated with `context`. Let `encrypted_hrr_inner` denote the
+extension payload. The client computes
+
+~~~
+    HelloRetryRequestInner = Open(hrr_inner_key,
+                                  hrr_inner_nonce,
+                                  "", // Empty AAD
+                                  encrypted_hrr_inner)
+~~~
+
+If decryption fails, then the client aborts with a "TODO" alert.
+
+The second ClientHelloOuter contains no extensions except for the
+"encrypted_client_hello" extension with the second ClientHelloInner. The client
+uses HelloRetryRequestInner to construct ClientHelloInner per the rules of
+{{RFC8446}}. It then encodes the second ClientHelloInner as in
```

Comment:
```suggestion
"encrypted_client_hello" extension with the second ClientHelloInner. All fields of
the second ClientHelloOuter are set to the same values used in the first ClientHelloOuter.
The client then constructs ClientHelloInner per the rules of {{RFC8446}}, Section 4.1.4. 
It then encodes the second ClientHelloInner as in
```

Reusing fields from the first CH seems simplest, since the server accepts ECH anyway. We might also recommend that the random field be different, but I don't think it matters that much?

Reply:
(It only affects AAD for the second CH encryption, but that's already different since there are no extensions in this second CH, so 🤷 )

Reply:
This point was discussed here: https://github.com/tlswg/draft-ietf-tls-esni/issues/374#issuecomment-748522170. David prefers to send a bare-bones CHO2 to reduce implementation complexity.

Reply:
Right, but what does "bare-bones" mean? You can't *omit* these values, otherwise it's not a CH, right?

Reply:
A CH with only one extension is still a CH, syntactically speaking. I don't see any problem with doing so. In any case, this is one of the design points we need to escalate.

Reply:
Clarified that only the extensions field changes in CH02.